File: Src/ExpectationLib/ExpectationExistsPatternMatcher.cs
namespace ExpectationLib;

/// <summary>
/// A matcher that expects an observation that meets an expectation to exist
/// </summary>
public class ExpectationExistsPatternMatcher : IExpectatedObservationsPattern
{
    public StimuliProducesResponseExpectation StimuliProducesResponseExpectation { get; }
    public List<Observation> Observations { get; }

    public ExpectationExistsPatternMatcher(StimuliProducesResponseExpectation stimuliProducesResponseExpectation, List<Observation> observations)
    {
        StimuliProducesResponseExpectation = stimuliProducesResponseExpectation;
        Observations = observations;
    }

    public bool Match()
    {
        return Observations.Any(o => o.Stimulus.Sender.Id.Equals(StimuliProducesResponseExpectation.Stimuli.Sender.Id) &&
                                     o.Stimulus.Receiver.Id.Equals(StimuliProducesResponseExpectation.Stimuli.Receiver.Id) &&
                                     o.Response.Id.Equals(StimuliProducesResponseExpectation.Response.Id));
    }
}

File: Src/CppLibExpectation/ExpectationExistsPatternMatcher.cpp
#include "pch.h"
#include "ExpectationExistsPatternMatcher.h"

namespace ExpectationLib
{
	ExpectationExistsPatternMatcher::ExpectationExistsPatternMatcher(
		const std::shared_ptr<StimuliProducesResponseExpectation>& stimuliProducesResponseExpectation,
		const std::vector<std::shared_ptr<Observation>>& observations)
	{
		this->stimuliProducesResponseExpectation = stimuliProducesResponseExpectation;
		this->Observations = observations;
	}

	bool ExpectationExistsPatternMatcher::Match()
	{
		for (const auto& o : Observations) 
		{
			if (o->GetStimulus()->GetSender()->GetId() == stimuliProducesResponseExpectation->GetStimulus()->GetSender()->GetId() &&
				o->GetStimulus()->GetReceiver()->GetId() == stimuliProducesResponseExpectation->GetStimulus()->GetReceiver()->GetId() &&
				o->GetResponse()->GetId() == stimuliProducesResponseExpectation->GetResponse()->GetId())
			{
				return true;
			}
		}
		return false;
	}
}
